src/PoissonSolvers/FFTOpenPoissonSolver.hpp

```diff
@@ -1747,4 +1928,570 @@ namespace ippl {
+        y[1] = ippl::Index(size[1], 2 * size[1] - 1);
+        y[2] = ippl::Index(size[2]);
+
+        ippl::NDIndex<Dim> z;
+        z[0] = ippl::Index(size[0]);
+        z[1] = ippl::Index(size[1]);
+        z[2] = ippl::Index(size[2], 2 * size[2] - 1);
+
+        ippl::NDIndex<Dim> xy;
+        xy[0] = ippl::Index(size[0], 2 * size[0] - 1);
+        xy[1] = ippl::Index(size[1], 2 * size[1] - 1);
+        xy[2] = ippl::Index(size[2]);
+
+        ippl::NDIndex<Dim> xz;
+        xz[0] = ippl::Index(size[0], 2 * size[0] - 1);
+        xz[1] = ippl::Index(size[1]);
+        xz[2] = ippl::Index(size[2], 2 * size[2] - 1);
+
+        ippl::NDIndex<Dim> yz;
+        yz[0] = ippl::Index(size[0]);
+        yz[1] = ippl::Index(size[1], 2 * size[1] - 1);
+        yz[2] = ippl::Index(size[2], 2 * size[2] - 1);
+
+        ippl::NDIndex<Dim> xyz;
+        for (unsigned i = 0; i < Dim; i++) {
+            xyz[i] = ippl::Index(size[i], 2 * size[i] - 1);
+        }
+
+        // 2nd step: send
+        for (int i = 0; i < ranks; ++i) {
+            auto domain2 = lDomains2[i];
+
+            if (domain2.touches(none)) {
+                auto intersection = domain2.intersect(none);
+
+                if (ldom_g.touches(intersection)) {
+                    intersection = intersection.intersect(ldom_g);
+                    requests.resize(requests.size() + 1);
+
+                    Communicate::size_type nsends;
+                    pack(intersection, view_g, fd_m, nghost_g, ldom_g, nsends);
+
+                    buffer_type buf =
+                        ippl::Comm->getBuffer<memory_space, Trhs>(IPPL_VICO_SEND + i, nsends);
+
+                    int tag = VICO_SOLVER_TAG;
+
+                    ippl::Comm->isend(i, tag, fd_m, *buf, requests.back(), nsends);
+                    buf->resetWritePos();
+                }
+            }
+
+            if (domain2.touches(x)) {
+                auto intersection = domain2.intersect(x);
+                auto xdom         = ippl::Index((2 * size[0] - intersection[0].first()),
+                                                (2 * size[0] - intersection[0].last()), -1);
+
+                ippl::NDIndex<Dim> domain2n1;
+                domain2n1[0] = xdom;
+                domain2n1[1] = intersection[1];
+                domain2n1[2] = intersection[2];
+
+                if (ldom_g.touches(domain2n1)) {
+                    intersection = ldom_g.intersect(domain2n1);
+
+                    requests.resize(requests.size() + 1);
+
+                    Communicate::size_type nsends;
+                    pack(intersection, view_g, fd_m, nghost_g, ldom_g, nsends);
+
+                    buffer_type buf =
+                        ippl::Comm->getBuffer<memory_space, Trhs>(IPPL_VICO_SEND + 8 + i, nsends);
+
+                    int tag = VICO_SOLVER_TAG + 1;
+
+                    ippl::Comm->isend(i, tag, fd_m, *buf, requests.back(), nsends);
+                    buf->resetWritePos();
+                }
```

Comment:
Can't we put this body into a function that is called with the different arguments? It seems to appear several times.

Reply:
Working on it. Might need a bit of specialization since it changes a bit throughout the code

Reply:
Added functions `solver_recv` and `solver_send` in the last commit. Please have a look